doc/guides/dts/index.rst

```diff
@@ -346,42 +346,113 @@ Below is sample header content generated for this node, in
-from ``reg``. Other suffixes, like ``*_LABEL``, are generated directly from the
-property name.
+Node prefixes
+-------------
 
-The second macro (``DT_INST_0_NXP_KINETIS_SIM_BASE_ADDRESS``) is an alias for
-the first macro. The node identifier ``...INST_0_NXP_KINETIS_SIM_...`` means
-"the first node with compatible string ``nxp,kinetis-sim``.
+Macros generated from particular devicetree nodes start with ``DT_<node>``,
+where ``<node>`` identifies the node. Several prefixes get generated for nodes:
 
-Aliases are also generated from any properties in the ``/aliases`` node. Take
-the DTS fragment below as an example.
+``DT_<compatible>_<unit-address>``
+    The compatible string for the node followed by its unit address.
+
+    For nodes that don't have a unit address (no ``...@<address>`` in the node
+    name), the unit address of the parent node is used instead. If the parent
+    node has no unit address either, the name of the node is used for
+    ``<unit-address>``, as a fallback.
+
+    For the ``sim@40047000`` node above, the compatible string is
+    ``"nxp,kinetis-sim"``. This gets converted to ``NXP_KINETIS_SIM`` by
+    uppercasing and replacing non-alphanumeric characters with underscores.
+    Adding the unit address gives the node prefix
+    ``DT_NXP_KINETIS_SIM_40047000``.
+
+``DT_INST_<instance-no.>_<compatible>``
+    The compatible string for the node together with an instance number.
+
+    The instance number is a unique index among all enabled
+    (``status = "okay"``) nodes that have a particular compatible string,
+    counting from zero. For example, if there are two enabled nodes that have
+    ``compatible = "foo,uart"``, then these node prefixes get generated:
+
+    .. code:: none
+
+        DT_INST_0_FOO_UART
+        DT_INST_1_FOO_UART
+
+``DT_ALIAS_<alias>``
+    Generated from the names of any properties in the ``/aliases`` node that
+    point to the node.
+
+    For example, assume ``/aliases`` looks like this:
+
+    .. code-block:: none
+
+       aliases {
+               uart-1 = &uart1;
+       };
+
+    The alias name (``i2c-1``) is converted to ``I2C_1`` by uppercasing and
+    replacing non-alphanumeric characters with underscores, generating the node
+    prefix ``DT_ALIAS_UART_1`` for the ``&uart1`` node.
+
+    .. note::
+
+       Currently, an older deprecated ``DT_<compatible>_<alias>`` form is also
+       generated for aliases. For the example above, assuming the compatible
+       string for the ``&uart1`` node is ``"foo,uart"``, this gives
+       ``DT_FOO_UART_UART_1``.
+
+       Work is underway to replace this form with ``DT_ALIAS_*``.
+
+Property suffixes
+-----------------
+
+Macros for particular properties on nodes have the form
+``DT_<node>_<property>``, where ``<node>`` is one the node prefixes from the
+section above, and ``<property>`` identifies the property.
+
+For all properties except ``reg``, ``<property>`` is just the property name
```

Comment:
First off, thanks for taking the time to improve the documentation. The information contained here looks good.

However, I do have some concerns.

I think this sentence is oversimplifying things too much, and I find myself agreeing with @pabigot's comment from the dev review meeting today that perhaps we need an "overall goals" section written down for devicetree in Zephyr before I feel like I can review this in context.

I know there is a comment at the bottom that says "this isn't the whole story", but this statement saying "for all properties except reg, <etc.>" is not correct.

There are many node properties which are exceptions to this rule.

- array type properties (array and string-array at least) get per-element macros with `_<i>` suffixes too (though this is not consistent, since uint8-array gets an initializer but no per-element macros)
- properties with type phandle-array have their own set of macros that take over a page to describe, which in part relates closely to the privileged position of the `label` property
- enums in the bindings generate `DT_<NODE>_<PROPERTY>_ENUM` macros
- flash partitions are partially covered in flash_partitions.inc, but their "label" properties are also handled as a special case
- interrupt tree properties (like `interrupts`) are handled as a special case
- clock tree properties (like `clocks`) are also a special case
- properties related to SPI device chip select data (cs-gpios properties) are a special case

Given that, I wonder where this is going:

1. Is this PR meant to be a step towards telling something closer to the whole story?
2. Are we always going to have a "RTFS" (read the fine source) comment left at the bottom?
3. Is that still up in the air?

I think getting some more context about where Zephyr is going with DT would help clear this up for me.

Thanks.

Reply:
The plan is just to document a bunch of cryptic stuff that's currently undocumented, to make the macros look less like a bunch of random stuff shoved together with underscores. More high-level documentation might be nice at some point too, but first the low-level stuff needs to be understandable.

I'll try not to lie anywhere.

Reply:
> More high-level documentation might be nice at some point too, but first the low-level stuff needs to be understandable.

If this means "it's a step towards telling something closer to the whole story" (item 1. in the list above) by documenting the current state of the world, then OK.

But a high level view of what the destination is from the perspective of Zephyr's users is not a "nice" thing to have "at some point" IMO. It's a critical step 1.